clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -51,7 +54,13 @@
     private final Selectable selector;
     
     private final MetadataUpdater metadataUpdater;
-
+    
+    /* a list of nodes we've connected to in the past */
+    private final List<Integer> nodesEverSeen;
+    private final Map<Integer, Node> nodesEverSeenById;
```

Comment:
Thanks @guozhangwang . Using a LinkedHashMap would allow us to get rid of one variable (nodesEverSeen), however it would 1) slightly complicate the code in leastLoadedNode that gets a random starting element from the list and iterating forward from that element in leastLoadedNode (inefficient iterating and wrapping around) and 2) the real cost of the LinkedHashMap is similar if not larger than what we currently have due to use of linked list. I'm leaning towards keeping what we have unless you feel strongly about this.